File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
def solution(numbers, hand):
    answer = ''
    thumbs = {"L": [3, 0], "R": [3, 2]}
    
    for number in numbers:
        row = (number - 1) // 3
        column = (number - 1) % 3
        if number == 0:
            row = 3
            column = 1
        
        if column == 0:
            thumbs["L"] = [row, column]
            answer += "L"
        elif column == 2:
            thumbs["R"] = [row, column]
            answer += "R"
        else:
            L_R, L_C = thumbs["L"]
            R_R, R_C = thumbs["R"]
            
            distance_L = abs(row - L_R) + abs(column - L_C)
            distance_R = abs(row - R_R) + abs(column - R_C)
            if distance_L > distance_R:
                thumbs["R"] = [row, column]
                answer += "R"
            elif distance_L < distance_R:
                thumbs["L"] = [row, column]
                answer += "L"
            else:
                if hand == "right":
                    thumbs["R"] = [row, column]
                    answer += "R"
                else:
                    thumbs["L"] = [row, column]
                    answer += "L"
    return answer

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.js
function solution(numbers, hand) {
    var answer = '';
    const thumbs = {left: [3, 0], right: [3, 2]}
    
    function pushPad(type, position) {
        switch (type) {
            case "L":
                answer += "L"
                thumbs.left = position
                break
            case "R":
                answer += "R"
                thumbs.right = position
                break
        }
    }
    
    for (let i = 0; i < numbers.length; i++) {
        const key = numbers[i]
        let position = [-1, -1]
        if (key === 0) position = [3, 1]
        else {
            const row = Math.floor((key - 1) / 3)
            const col = (key - 1) % 3
            position = [row, col]
        }
        
        switch (position[1]){
            case 0:
                {
                    pushPad("L", position)
                    break
                }

            case 2:
                {
                    pushPad("R", position)
                    break
                }
            case 1:
                {
                    const leftD = Math.abs(thumbs.left[0] - position[0]) + Math.abs(thumbs.left[1] - position[1])
                    const rightD = Math.abs(thumbs.right[0] - position[0]) + Math.abs(thumbs.right[1] - position[1])
                    if (leftD > rightD) {
                        pushPad("R", position)
                    } else if (leftD < rightD) {
                        pushPad("L", position)
                    } else {
                        if (hand === "right") {
                            pushPad("R", position)
                        } else {
                            pushPad("L", position)
                        }
                    }
                    break
                }
        }
    }
    return answer;
}